Public/Src/Cache/ContentStore/App/DistributedService.cs

```diff
@@ -118,9 +118,40 @@ public void OnTeardownCompleted()
             {
             }
 
-            public Task<Dictionary<string, string>> RetrieveKeyVaultSecretsAsync(List<string> secrets, CancellationToken token)
+            public Task<Dictionary<string, Credentials>> RetrieveKeyVaultSecretsAsync(List<RetrieveSecretsRequest> requests, CancellationToken token)
             {
-                return Task.FromResult(secrets.ToDictionary(s => GetSecretStoreValue(s)));
+                var result = new Dictionary<string, Credentials>();
+
+                foreach (var request in requests)
+                {
+                    Credentials credentials = null;
+
+                    var secret = GetSecretStoreValue(request.Name);
+                    if (string.IsNullOrEmpty(secret))
+                    {
+                        // Environment variables are null by default. Skip if that's the case.
+                        continue;
+                    }
+
+                    switch (request.Kind)
+                    {
+                        case CredentialsKind.AzureBlobPlainText:
+                            credentials = new AzureBlobStorageCredentials(secret);
+                            break;
+                        case CredentialsKind.EventHubPlainText:

```

Comment:
Do we really need to separate the sources? For instance, why not to use the same set of secret kinds that css provides? Like `PlainText`, `Sas` etc.

Reply:
It is required to create the corresponding object for the authentication method, which means that the other side needs to know what is needed in each case; it is also helpful to validate that the other side is returning the expected objects and not others. The alternative is to keep two separate enums, one { EventHub, Redis, AzureBlob } and another { SAS, PAT }

---
In reply to: [303139192](https://github.com/microsoft/BuildXL/pull/587#discussion_r303139192) [](ancestors = 303139192)